src/decision/justification_strategy.cpp

```diff
@@ -0,0 +1,638 @@
+  TNode atom = pol ? n : n[0];
+  Assert(atom.getKind() != NOT);
+  // check if we have already determined the value
+  // notice that d_justified may contain nodes that are not assigned SAT values,
+  // since this class infers when the value of nodes can be determined.
+  context::CDInsertHashMap<Node, SatValue, NodeHashFunction>::const_iterator
+      jit = d_justified.find(atom);
+  if (jit != d_justified.end())
+  {
+    return pol ? jit->second : invertValue(jit->second);
+  }
+  // Notice that looking up values for non-theory atoms may lead to
+  // an incomplete strategy where a formula is asserted but not justified
+  // via its theory literal subterms. This is the case because the justification
+  // heuristic is not the only source of decisions, as the theory may request
+  // them.
+  if (isTheoryAtom(atom))
+  {
+    SatLiteral nsl = d_cnfStream->getLiteral(atom);
+    prop::SatValue val = d_satSolver->value(nsl);
+    if (val != SAT_VALUE_UNKNOWN)
+    {
+      // this is the moment where we realize a skolem definition is relevant,
+      // add now.
+      // NOTE: could call notifyJustified(atom) here
+      d_justified.insert(atom, val);
+      return pol ? val : invertValue(val);
+    }
+  }
+  return SAT_VALUE_UNKNOWN;
+}
+
+bool JustificationStrategy::isDone() { return !refreshCurrentAssertion(); }
+
+void JustificationStrategy::addAssertion(TNode assertion)
+{
+  std::vector<TNode> toProcess;
+  toProcess.push_back(assertion);
+  insertToAssertionList(toProcess, false);
+}
+
+void JustificationStrategy::addSkolemDefinition(TNode lem, TNode skolem)
+{
+  if (d_jhSkRlvMode == options::JutificationSkolemRlvMode::ALWAYS)
+  {
+    // just add to main assertions list
+    std::vector<TNode> toProcess;
+    toProcess.push_back(lem);
+    insertToAssertionList(toProcess, false);
+  }
+}
+
+void JustificationStrategy::notifyAsserted(TNode n)
+{
+  if (d_jhSkRlvMode == options::JutificationSkolemRlvMode::ASSERT)
+  {
+    // assertion processed makes all skolems in assertion active,
+    // which triggers their definitions to becoming relevant
+    std::vector<TNode> defs;
+    d_skdm->notifyAsserted(n, defs, true);
+    insertToAssertionList(defs, true);
+  }
+  // NOTE: can update tracking triggers, pop stack to where a child implied
+  // that a node on the current stack is justified.
+}
+
+void JustificationStrategy::insertToAssertionList(std::vector<TNode>& toProcess,
+                                                  bool useSkolemList)
+{
+  AssertionList& al = useSkolemList ? d_skolemAssertions : d_assertions;
+  // always miniscope AND and negated OR immediately
+  size_t index = 0;
+  // must keep some intermediate nodes below around for ref counting
+  std::vector<Node> keep;
+  while (index < toProcess.size())
+  {
+    TNode curr = toProcess[index];
+    bool pol = curr.getKind() != NOT;
+    TNode currAtom = pol ? curr : curr[0];
+    index++;
+    Kind k = currAtom.getKind();
+    if (k == AND && pol)
```

Comment:
This is for simplicity.  `toProcess` is a temporary list that we are iterating over.  There isn't really a reason to erase it from the list since we've iterated past it.  Otherwise, this would require additional work to erase it.